R/odin.R

```diff
@@ -44,8 +24,90 @@ odin_js_validate <- function(code) {
       msg <- result$error$message
       line <- integer(0)
     }
+    return(list(valid = scalar(FALSE),
+                error = list(message = scalar(msg), line = line)))
+  }
+
+  dat <- result$result
+  messages <- lapply(result$messages, function(m)
+    list(message = scalar(m$msg), line = m$line))
+
+  variables <- c(names(dat$data$variable$contents),
+                 names(dat$data$output$contents))
+
+  process_user <- function(nm) {
+    x <- dat$equations[[nm]]$user
+    list(name = scalar(nm),
+         default = scalar(x$default %||% NA),
+         min = scalar(x$min %||% NA),
+         max = scalar(x$max %||% NA),
+         is_integer = scalar(x$integer %||% FALSE),
+         rank = scalar(dat$data$elements[[nm]]$rank))
+  }
+  parameters <- lapply(names(dat$user), process_user)
+
+  list(valid = scalar(TRUE),
+       metadata = list(
+         variables = variables,
+         parameters = parameters,
+         messages = messages))
+}
 
-    list(valid = scalar(FALSE),
-         error = list(message = scalar(msg), line = line))
+
+check_requirements <- function(result, requirements) {
+  if (is.null(requirements)) {
+    requirements <- list(timeType = "continuous")
+  }
+
+  ## This is the error we'd get if the server has asked that we check
+  ## we provide a discrete time model, regardless of what the
+  ## user-provided code is. We can't do this yet. This error should
+  ## never be surfaced in the app unless we've deployed incompatible
+  ## versions.
+  if (requirements$timeType != "continuous") {
+    msg <- "Only continuous time models currently supported"
+    return(odin_validate_error_value(msg, integer(0)))
+  }
+
+  dat <- result$result
+
+  if (dat$features$discrete) {
+    ## Once discrete time models are supported at all, we will also need
+    ## to check that stochastic models do not use output (or
+    ## interpolation?) as this is not supported in dust, even though it
+    ## still is in odin itself.
+    msg <- "Expected a model continuous time model (using deriv, not update)"
+    vars <- names(dat$data$variable$contents)
+    line <- lapply(dat$equations, function(el) {
+      if (el$lhs %in% vars && el$name %in% dat$components$rhs)
+        el$source else NULL
+    })
```

Comment:
This is finding the line in the code where the error originates?

Reply:
yes, that's right. won't be that much use until we have the line numbering sorted though. This will highlight every line that contains a `update(name) <- equation` line instead of a `deriv(name) <- equation` one